### ISSUE TITLE
 Error tailing oplog entries: Can't use 'local' database through mongos

### ISSUE THREAD

**Issue by lolgolflol** [rwynn/monstache#207]
INFO 2019/04/18 06:23:43 Started monstache version 4.16.1
INFO 2019/04/18 06:23:43 Successfully connected to MongoDB version 4.0.9
INFO 2019/04/18 06:23:43 Successfully connected to Elasticsearch version 6.3.2
INFO 2019/04/18 06:23:43 Listening for events
ERROR 2019/04/18 06:23:43 Error tailing oplog entries: Can't use 'local' database through mongos

In part 

```toml
# connect to MongoDB using the following URL
mongo-url = "mongodb://localhost:27017"
# connect to the Elasticsearch REST API at the following node URLs
elasticsearch-urls = ["http://localhost:9200"]
```
what mongo-url should I input ( mongos (routing) or mongod as primary shard )? but I try to use the both, I still get the same error.

**Comment by lolgolflol** [comment]
#105  I found this solve

**Comment by rwynn** [comment]
Hi @lolgolflol glad your found a solution.  Just wanted to clarify some things.  

For backwards compatibility monstache defaults to tail the oplog collection (`local.oplog.rs`) directly.  This was the only option for listening for changes prior to MongoDB version 3.6.  In MongoDB 3.6 a new API was added such that changes can be watched without access to the oplog.  Monstache supports this new API but only if you configure it with the option `change-stream-namespaces`.  Since you are using MongoDB 4.0.9 you can enable this option.  The following will use the API to watch changes on the entire deployment:

```
change-stream-namespaces = [ "" ]
```

Change streams work through a mongos connection, so when you enable change streams you should not set `mongo-config-url`.   With change streams enabled you should only set `mongo-url` to the URL of you mongos or mongod.  

The `mongo-config-url` should only be used if you do not configure change streams (maybe because you are using MongoDB prior to 3.6) and you have a clustered MongoDB.  Accessing the oplog is only possible on a mongod connection not through a mongos connection.  Therefore, what monstache requires in this case is to connect to the MongoDB config server at the `mongo-config-url` and collect the URLs of all the shards and connect to them directly such that it can access the oplog.  

So in summary, if you have a clustered MongoDB then you should always put the mongos URL in `mongo-url`.  Next, if you have MongoDB 3.6+ you should configure `change-stream-namespaces` and not configure `mongo-config-url`.  If you have MongoDB prior to 3.6, you should configure `mongo-config-url` and not configure `change-stream-namespaces`.  

If you have MongoDB 3.6+ you don't **need** to configure `change-stream-namespaces` but this is not recommended since the change streams API is superior in many ways.

**Comment by lolgolflol** [comment]
Thank you very much, It helped me a lot and new knowledge for me.
